File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class Solution {
    public int solution(string[] friends, string[] gifts) {
         int answer = 0;

 int friendsCount = friends.Length;
 int[] giftIndex = new int[friendsCount];
 int[,] giftHistory= new int[friendsCount, friendsCount];
 int[] giftCount = new int[friendsCount];

 foreach(string gift in gifts)
 {
     string[] history = gift.Split(' ');

     int give = Array.IndexOf(friends, history[0]);
     int take = Array.IndexOf(friends, history[1]);

     giftIndex[give]++;
     giftIndex[take]--;
     giftHistory[give, take]++;
 }

 for (int i = 0; i < friendsCount; i++)
 {
     for (int j = 0; j < friendsCount; j++)
     {
         if (i == j) continue;

         if (giftHistory[i, j] > giftHistory[j, i] || (giftIndex[i] > giftIndex[j] && giftHistory[i, j] == giftHistory[j, i]))
         {
             giftCount[i]++;
         }
     }
 }

 answer = giftCount.Max();

 return answer;
    }
}

File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.py
def solution(friends, gifts):
    answer = 0
    friends_length = len(friends)

    give_and_take_matrix = [[0] * friends_length for _ in range(friends_length)]
    gift_index = [0] * friends_length
    gift_count = [0] * friends_length

    friends_dict = {name: index for index, name in enumerate(friends)}

    for give_and_take_history in gifts:
        split_give_and_take = give_and_take_history.split(" ")
        givePerson = friends_dict[split_give_and_take[0]]
        takePerson = friends_dict[split_give_and_take[1]]

        # 선물지수
        gift_index[givePerson] += 1
        gift_index[takePerson] -= 1

        # 두 사람 간 선물 내역
        give_and_take_matrix[givePerson][takePerson] += 1

    for i in range(0, friends_length):
        for j in range(0, friends_length):
            if i == j:
                continue

            if give_and_take_matrix[i][j] > give_and_take_matrix[j][i] or (
                give_and_take_matrix[i][j] == give_and_take_matrix[j][i]
                and gift_index[i] > gift_index[j]
            ):
                gift_count[i] += 1

    answer = max(gift_count)

    return answer